design/host-profiles.md

```diff
@@ -0,0 +1,2008 @@
+to provide a label similar to an instance flavor in a traditional
+cloud environment. Host profiles are configurable through the
+Kubernetes API, allowing cluster administrators to expand the known
+hardware platforms and control how they are used in the cluster.
+
+## Motivation
+
+We want to limit the provisioning operations to known hardware
+platforms, while making the set of known platforms easy to expand over
+time. We do not want to hard-code the definitions, because that makes
+it more difficult to configure development or test environments that
+may not match what an expected production environment looks like.
+
+### Goals
+
+- Provide a place for us to store known hardware profiles.
+- Provide a way to match known profiles to actual hardware.
+- Provide a way to control how that hardware is used as part of a
+  MachineSet.
+
+### Non-Goals
+
+This goal is about examining the hardware present in the host. It does
+not consider whether that hardware is returning fault codes or
+otherwise not working properly. Dealing with faulty hardware is
+covered under a separate design.
+
+For now, we only want to do simple matching based on data, and we do
+not want a complex "language" to describe matching rules. It is enough
+for hosts to look like the hardware described in the profile.
+
+This design does not cover using the profile to control how software
+is deployed to the host (specifying networking access, storage use,
+etc.). That may come in a later design.
+
+## Proposal
+
+A new CRD, BareMetalHostProfile, is defined with Spec fields for all
+of the relevant hardware considerations. Ideally this would reuse the
+hardware specification data structure in the BareMetalHost itself, or
+be a different structure that summarizes that structure (for example,
+providing numCPUs and cpuGHz, fields but not a list of CPUs and their
+individual speeds).
+
+The Status portion of the BareMetalHostProfile will report all of the
+BareMetalHosts and MachineSets associated with the profile.
+
+*We need more detail about exactly what sort of hardware matching we
+care about.*
+
+When a profile is created, the profile operator causes all host
+resources labeled as having an unknown profile to be reconciled to see
+if they match the new profile. This means the order of creation for
+hosts and profiles does not matter, because the match will happen
+eventually.
+
+When a profile is updated, the profile operator causes all host
+resources associated with that profile to be reconciled to see if they
+still match the updated profile.
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if the design is
+implemented.  Include as much detail as possible so that people can
+understand the "how" of the system.  The goal here is to make this
+feel real for users without getting bogged down.
+
+#### Controlling Types of Hosts Added to the Cluster
+
+The machine actuator in the cluster-api-provider-baremetal repository
```

Comment:
This means to me that the concept of host-profiles is a bit underwhelming for what I would have hoped.  So far it seems like the primary motivation is for scheduling purposes.  When I think of host profiles as mentioned above, I also imagine not only validation as mentioned above, but also a way to de-duplicate commonality in like-machines.  For instance, specifying the cloud-init definition at a profile level rather than a baremetalnode level.

Reply:
I think you're combining some of the role of the provider spec from the MachineClass or MachineDeployment with the profile. Maybe "profile" is a bad term to use for this object.